commit 57f0988706fec1b8dbc3fe00965828a47e2235a1
Author: Jason Gunthorpe <jgg@ziepe.ca>
Date:   Tue Nov 29 16:29:42 2022 -0400

    iommufd: Add a selftest
    
    Cover the essential functionality of the iommufd with a directed test from
    userspace. This aims to achieve reasonable functional coverage using the
    in-kernel self test framework.
    
    A second test does a failure injection sweep of the success paths to study
    error unwind behaviors.
    
    This allows achieving high coverage of the corner cases in pages.c.
    
    The selftest requires CONFIG_IOMMUFD_TEST to be enabled, and several huge
    pages which may require:
    
      echo 4 > /proc/sys/vm/nr_hugepages
    
    Link: https://lore.kernel.org/r/19-v6-a196d26f289e+11787-iommufd_jgg@nvidia.com
    Tested-by: Nicolin Chen <nicolinc@nvidia.com>
    Tested-by: Matthew Rosato <mjrosato@linux.ibm.com> # s390
    Tested-by: Yi Liu <yi.l.liu@intel.com>
    Tested-by: Eric Auger <eric.auger@redhat.com> # aarch64
    Signed-off-by: Nicolin Chen <nicolinc@nvidia.com>
    Signed-off-by: Yi Liu <yi.l.liu@intel.com>
    Signed-off-by: Jason Gunthorpe <jgg@nvidia.com>

diff --git a/tools/testing/selftests/iommu/Makefile b/tools/testing/selftests/iommu/Makefile
new file mode 100644
index 000000000000..7cb74d26f141
--- /dev/null
+++ b/tools/testing/selftests/iommu/Makefile
@@ -0,0 +1,12 @@
+# SPDX-License-Identifier: GPL-2.0-only
+CFLAGS += -Wall -O2 -Wno-unused-function
+CFLAGS += -I../../../../include/uapi/
+CFLAGS += -I../../../../include/
+
+CFLAGS += -D_GNU_SOURCE
+
+TEST_GEN_PROGS :=
+TEST_GEN_PROGS += iommufd
+TEST_GEN_PROGS += iommufd_fail_nth
+
+include ../lib.mk